files/puma_metrics

```diff
@@ -0,0 +1,161 @@
+#!/opt/puppetlabs/puppet/bin/ruby
+
+require "net/https"
+require "json"
+require "uri"
+require 'time'
+require 'optparse'
+require 'yaml'
+
+options = {}
+OptionParser.new do |opts|
+  opts.banner = "Usage: puma_metrics [options]"
+
+  opts.on('-p', '--[no-]print', 'Print to stdout') { |p| options[:print] = p }
+  opts.on('-m [TYPE]', '--metrics_type [TYPE]', 'Type of metric to collect') { |v| options[:metrics_type] = v }
+  opts.on('-o [DIR]', '--output-dir [DIR]', 'Directory to save output to') { |o| options[:output_dir] = o }
+  opts.on('--metrics_port [PORT]', 'The port the metrics service runs on') { |port| options[:metrics_port] = port }
+  opts.on('--[no-]ssl', 'Whether or not to use SSL when gather metrics') { |ssl| options[:ssl] = ssl }
+end.parse!
+
+if options[:metrics_type].nil? then
+  STDERR.puts '--metrics_type (-m) is a required argument'
+  exit 1
+end
+
+METRICS_TYPE = options[:metrics_type]
+config = YAML.load_file(File.join(File.dirname(File.expand_path(__FILE__)),"#{METRICS_TYPE}_config.yaml"))
+
+def coalesce ( opt1, opt2, default=nil )
+  if opt1.nil? then
+    if opt2.nil? then
+      default
+    else
+      opt2
+    end
+  else
+    opt1
+  end
+end
+
+OUTPUT_DIR = options[:output_dir]
+HOSTS      = config['hosts']
+PORT       = coalesce(options[:metrics_port], config['metrics_port'])
+METRICS    = config['additional_metrics']
+PE_VERSION = config['pe_version']
+SSL        = coalesce(options[:ssl], config['ssl'], true)
+EXCLUDES   = config['excludes']
+
+if SSL
+  CERTNAME = `puppet config print certname`.chomp
+  SSLDIR   = `puppet config print ssldir`.chomp
+end
+
+$error_array = []
+
+# Change to reflect the version of PE with Puma metrics endpoints.
+exit if Gem::Version.new(PE_VERSION) < Gem::Version.new('2019.4.0')
```

Comment:
How does this behave on OSP?

Reply:
This module has a couple of other references to `pe_version` that we need to audit for OSP support:

https://github.com/puppetlabs/puppetlabs-puppet_metrics_collector/search?q=pe_version&unscoped_q=pe_version

I will update the version to reflect that the endpoints are available in 2019.3.0 ...

Reply:
In `tune` I check Puppet version (as I'm worried that /opt/puppetlabs/server/pe_version may not exist during installation, or be out of date during an upgrade):

https://github.com/tkishel/pe_tune/blob/master/lib/puppet_x/puppetlabs/tune.rb#L979

I correlate those Puppet to PE versions via:

https://puppet.com/docs/pe/latest/component_versions_in_recent_pe_releases.html

Maybe we could/should do the same here?

Reply:
Makes sense to me. I tagged this in the wrong place. Iif `$facts['pe_server_version']` is undef ideally it would not fail with `undef` error, not necessarily that it is compatible with OSP. https://github.com/puppetlabs/puppetlabs-puppet_metrics_collector/pull/23 added a simple check on the fact when using it in other classes. It probably doesn't matter much since ace and bolt would not be services in OSP, but they are automatically included in the `init.pp` so it could cause some noise on the cron output.

Reply:
I'm happy to review all of those references, after this is merged.

Also would you be willing to create graphs for these metrics (and Orchestrator metrics, which we already collect but do not graph) in the Puppet Metrics Dashboard? I have no Grafana skillz :)

Reply:
Sounds like a plan to me. This is an amazing addition to the data. I need to create some for the SAR metrics and will do these as well.

Reply:
😊

I'm hoping these metrics (and Orchestrator metrics) will help debug a variety of issues.

Reply:
I have some new graphs here: 
http://10.234.1.31:3000/d/R3mUROsZz/archive-puppetserver-performance?orgId=1&from=1579707827810&to=1579794857230&var-server=763a_rc4_PARGC_NMT_24hr&var-server=763b_rc4_G1_NMT_24hr&panelId=24
Just some we put together as part of the java11 investigation.  It will need reorganizing.  And there is some changed to the importer that need to be made to treat ace and bolt like the other services too.  Maybe we should start a doc with a graph wishlist?   
I also think I need to change some things about the sar metrics... puppetserver reports as 100%, but top says 185%... So i don't think I am handling multiple cores properly.